### PR TITLE
Iio demo

### PR DESCRIPTION
**Pull request by cristiantpop**
iio_demo: Add iio demo module and application

Registers a iio_demo_interfcae, independent of a physical device and should
be used as reference, when creating a new iio application for a new iio
device.

Signed-off-by: Cristian Pop <cristian.pop@analog.com>

### CODE REVIEW

File: iio/iio_demo/iio_demo.c [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+		.get_xml = iio_demo_get_xml,
+		.transfer_dev_to_mem = iio_demo_transfer_dev_to_mem,
+		.read_data = iio_demo_read_dev,
+		.transfer_mem_to_dev = iio_demo_transfer_mem_to_dev,
+		.write_data = iio_demo_write_dev,
+	};
+
+	status = iio_register(iio_interface);
+	if (status < 0)
+		goto error_free_iio_demo_delete_dev;
```
Comment (mchindri) [inline iio/iio_demo/iio_demo.c]: It this fails you also have to free iio_interface

**Comment by cristiantpop** [comment]
V2:

- Add parentheses to macros;

- Add one line comments;

- Free iio_interface if "iio_register" fails.

File: iio/iio_types.h [hunk 2]
```diff
@@ -45,6 +45,7 @@
 
 #include <stdbool.h>
 #include <stdint.h>
+#include <stdio.h>
```
Comment (buha) [inline iio/iio_types.h]: What is it needed for ? I don't see any other change in iio_types.h, so I suspect that it is needed only for something in a .c file, in which case it is more suitable to move it to that .c file

Comment (cristiantpop) [reply iio/iio_types.h]: Actually the file needed here is  "types.h".

**Comment by cristiantpop** [comment]
V3:

- Include missing "types.h" header file, for "ssize_t" type.

File: iio/iio_types.h [hunk 3]
```diff
@@ -45,6 +45,7 @@
```
Comment (adrimbarean) [inline iio/iio_types.h]: Could you explain in the commit body why it is needed?

Comment (cristiantpop) [reply iio/iio_types.h]: Sure

**Comment by cristiantpop** [comment]
V4:

- Explain in the commit body why "types.h", needs to be included.

**Comment by cristiantpop** [comment]
V5:

- Add explicit casts to be compatible with C++ compilers

File: iio/iio_demo/iio_demo.c [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+	.ch_out = false,
+};
+
+static struct iio_channel *iio_demo_channels[] = {
+	&iio_demo_channel_voltage0_in,
+	NULL,
+};
+
+static struct iio_attribute *iio_demo_global_attributes[] = {
+	&iio_attr_demo,
```
Comment (dbogdan) [inline iio/iio_demo/iio_demo.c]: Let's create an independent iio_attribute for channels.

**Comment by cristiantpop** [comment]
V6:

- Create an independent iio_attribute for channels.

**Pull request merged**